File: src/main/java/com/budgetbuddy/service/WeekendOverspendingService.java
package com.budgetbuddy.service;

import com.budgetbuddy.model.Transaction;
import com.budgetbuddy.model.User;
import com.budgetbuddy.model.WeekendOverspending;
import com.budgetbuddy.repository.TransactionRepository;
import com.budgetbuddy.repository.WeekendOverspendingRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class WeekendOverspendingService {

    private static final Logger logger = LoggerFactory.getLogger(WeekendOverspendingService.class);

    @Autowired
    private TransactionRepository transactionRepository;

    @Autowired
    private WeekendOverspendingRepository weekendOverspendingRepository;

    /**
     * Check if a transaction category is an investment (should be excluded from expense calculations).
     * Investments are not considered expenses because they represent asset allocation, not consumption.
     */
    private boolean isInvestmentCategory(String category) {
        if (category == null || category.trim().isEmpty()) {
            return false;
        }
        String normalized = category.toLowerCase().trim();
        return normalized.equals("investments") || normalized.startsWith("investments /");
    }

    /**
     * Check if a transaction should be excluded from expense calculations.
     * A transaction is excluded if either its categoryName or predictedCategory is an investment.
     */
    private boolean isInvestmentTransaction(Transaction transaction) {
        return isInvestmentCategory(transaction.getCategoryName()) || 
               isInvestmentCategory(transaction.getPredictedCategory());
    }

    /**
     * Detect weekend overspending patterns
     */
    public List<WeekendOverspending> detectWeekendOverspending(User user) {
        logger.info("Detecting weekend overspending for user: {}", user.getId());

        YearMonth currentMonth = YearMonth.now();
        LocalDate monthStart = currentMonth.atDay(1);
        LocalDate monthEnd = currentMonth.atEndOfMonth();

        // Get current month transactions, excluding investments
        List<Transaction> transactions = transactionRepository.findAll()
            .stream()
            .filter(t -> t.getUser().getId().equals(user.getId()))
            .filter(t -> t.getAmount() != null && t.getAmount() < 0)
            .filter(t -> !isInvestmentTransaction(t))
            .filter(t -> !t.getDate().isBefore(monthStart) && !t.getDate().isAfter(monthEnd))
            .collect(Collectors.toList());

        if (transactions.isEmpty()) {
            return Collections.emptyList();
        }

        // Group by category
        Map<String, List<Transaction>> byCategory = transactions.stream()
            .filter(t -> t.getPredictedCategory() != null)
            .collect(Collectors.groupingBy(Transaction::getPredictedCategory));

        List<WeekendOverspending> results = new ArrayList<>();

        for (Map.Entry<String, List<Transaction>> entry : byCategory.entrySet()) {
            String category = entry.getKey();
            List<Transaction> categoryTxs = entry.getValue();

            // Separate weekend (Sat, Sun) and weekday (Mon-Fri)
            List<Transaction> weekendTxs = categoryTxs.stream()
                .filter(t -> {
                    DayOfWeek day = t.getDate().getDayOfWeek();
                    return day == DayOfWeek.SATURDAY || day == DayOfWeek.SUNDAY;
                })
                .collect(Collectors.toList());

            List<Transaction> weekdayTxs = categoryTxs.stream()
                .filter(t -> {
                    DayOfWeek day = t.getDate().getDayOfWeek();
                    return day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY;
                })
                .collect(Collectors.toList());

            if (weekendTxs.isEmpty() || weekdayTxs.isEmpty()) {
                continue;
            }

            // Calculate averages
            double weekendAvg = weekendTxs.stream()
                .mapToDouble(t -> Math.abs(t.getAmount()))
                .average()
                .orElse(0.0);

            double weekdayAvg = weekdayTxs.stream()
                .mapToDouble(t -> Math.abs(t.getAmount()))
                .average()
                .orElse(0.0);

            if (weekdayAvg == 0) {
                continue;
            }

            double ratio = weekendAvg / weekdayAvg;
            double percentageIncrease = (ratio - 1.0) * 100;

            // Only flag if weekend spending is 30%+ higher than weekday
            if (ratio > 1.3) {
                WeekendOverspending overspending = new WeekendOverspending();
                overspending.setUser(user);
                overspending.setCategory(category);
                overspending.setWeekendAvg(weekendAvg);
                overspending.setWeekendSpending(weekendTxs.stream()
                    .mapToDouble(t -> Math.abs(t.getAmount()))
                    .sum());
                overspending.setWeekdayAvg(weekdayAvg);
                overspending.setWeekdaySpending(weekdayTxs.stream()
                    .mapToDouble(t -> Math.abs(t.getAmount()))
                    .sum());
                overspending.setRatio(ratio);
                overspending.setPercentageIncrease(percentageIncrease);
                overspending.setMonth(currentMonth);
                overspending.setYear(currentMonth.getYear());
                overspending.setTrend(determineTrend(user, category));
                overspending.setAlertLevel(determineAlertLevel(ratio));
                overspending.setIsActive(true);

                results.add(overspending);
            }
        }

        // Save results
        for (WeekendOverspending result : results) {
            weekendOverspendingRepository.save(result);
        }

        logger.info("Detected {} weekend overspending patterns for user {}", results.size(), user.getId());
        return results;
    }

    /**
     * Determine trend (increasing, decreasing, stable)
     */
    private WeekendOverspending.Trend determineTrend(User user, String category) {
        // Get last 3 months of weekend spending, excluding investments
        LocalDate threeMonthsAgo = LocalDate.now().minusMonths(3);
        List<Transaction> transactions = transactionRepository.findAll()
            .stream()
            .filter(t -> t.getUser().getId().equals(user.getId()))
            .filter(t -> t.getPredictedCategory() != null && t.getPredictedCategory().equals(category))
            .filter(t -> t.getAmount() != null && t.getAmount() < 0)
            .filter(t -> !isInvestmentTransaction(t))
            .filter(t -> t.getDate().isAfter(threeMonthsAgo))
            .collect(Collectors.toList());

        Map<YearMonth, Double> monthlyWeekendSpending = transactions.stream()
            .filter(t -> {
                DayOfWeek day = t.getDate().getDayOfWeek();
                return day == DayOfWeek.SATURDAY || day == DayOfWeek.SUNDAY;
            })
            .collect(Collectors.groupingBy(
                t -> YearMonth.from(t.getDate()),
                Collectors.summingDouble(t -> Math.abs(t.getAmount()))
            ));

        if (monthlyWeekendSpending.size() < 2) {
            return WeekendOverspending.Trend.STABLE;
        }

        List<Double> amounts = monthlyWeekendSpending.entrySet().stream()
            .sorted(Map.Entry.comparingByKey())
            .map(Map.Entry::getValue)
            .collect(Collectors.toList());

        double first = amounts.get(0);
        double last = amounts.get(amounts.size() - 1);

        if (last > first * 1.1) {
            return WeekendOverspending.Trend.INCREASING;
        } else if (last < first * 0.9) {
            return WeekendOverspending.Trend.DECREASING;
        }

        return WeekendOverspending.Trend.STABLE;
    }

    /**
     * Determine alert level based on ratio
     */
    private WeekendOverspending.AlertLevel determineAlertLevel(double ratio) {
        if (ratio > 1.5) {
            return WeekendOverspending.AlertLevel.HIGH;
        } else if (ratio > 1.3) {
            return WeekendOverspending.AlertLevel.MEDIUM;
        }
        return WeekendOverspending.AlertLevel.LOW;
    }

    /**
     * Get active weekend overspending alerts
     */
    public List<WeekendOverspending> getActiveAlerts(User user) {
        return weekendOverspendingRepository.findByUserAndIsActiveTrue(user);
    }
}

File: mobile-version/app/src/main/java/com/budgetbuddy/mobile/service/WeekendOverspendingService.kt
package com.budgetbuddy.mobile.service

import com.budgetbuddy.mobile.data.dao.TransactionDao
import com.budgetbuddy.mobile.data.dao.WeekendOverspendingDao
import com.budgetbuddy.mobile.data.model.WeekendOverspending
import kotlinx.coroutines.flow.first
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.YearMonth
import kotlin.math.abs

class WeekendOverspendingService(
    private val transactionDao: TransactionDao,
    private val weekendDao: WeekendOverspendingDao
) {
    
    /**
     * Check if a transaction category is an investment
     */
    private fun isInvestmentCategory(category: String?): Boolean {
        if (category == null || category.trim().isEmpty()) {
            return false
        }
        val normalized = category.lowercase().trim()
        return normalized == "investments" || normalized.startsWith("investments /")
    }
    
    /**
     * Check if a transaction is an investment
     */
    private fun isInvestmentTransaction(transaction: com.budgetbuddy.mobile.data.model.Transaction): Boolean {
        return isInvestmentCategory(transaction.categoryName) || 
               isInvestmentCategory(transaction.predictedCategory)
    }
    
    /**
     * Determine trend based on historical data
     */
    private fun determineTrend(
        currentRatio: Double,
        historicalRatios: List<Double>
    ): WeekendOverspending.Trend {
        if (historicalRatios.isEmpty()) {
            return WeekendOverspending.Trend.STABLE
        }
        
        val avgHistorical = historicalRatios.average()
        
        return when {
            currentRatio > avgHistorical * 1.1 -> WeekendOverspending.Trend.INCREASING
            currentRatio < avgHistorical * 0.9 -> WeekendOverspending.Trend.DECREASING
            else -> WeekendOverspending.Trend.STABLE
        }
    }
    
    /**
     * Determine alert level based on ratio
     */
    private fun determineAlertLevel(ratio: Double): WeekendOverspending.AlertLevel {
        return when {
            ratio > 1.5 -> WeekendOverspending.AlertLevel.HIGH
            ratio > 1.3 -> WeekendOverspending.AlertLevel.MEDIUM
            else -> WeekendOverspending.AlertLevel.LOW
        }
    }
    
    /**
     * Detect weekend overspending patterns
     */
    suspend fun detectWeekendOverspending(userId: Long): List<WeekendOverspending> {
        val currentMonth = YearMonth.now()
        val monthStart = currentMonth.atDay(1)
        val monthEnd = currentMonth.atEndOfMonth()
        
        // Get current month transactions, excluding investments
        val allTransactions = transactionDao.getTransactionsByUser(userId).first()
        val monthStartDate = currentMonth.atDay(1)
        val monthEndDate = currentMonth.atEndOfMonth()
        val transactions = allTransactions
            .filter { it.userId == userId }
            .filter { it.amount < 0 }
            .filter { !isInvestmentTransaction(it) }
            .filter { it.date.isAfter(monthStartDate.minusDays(1)) && 
                      it.date.isBefore(monthEndDate.plusDays(1)) }
        
        if (transactions.isEmpty()) {
            return emptyList()
        }
        
        // Group by category
        val byCategory = transactions
            .filter { it.predictedCategory != null }
            .groupBy { it.predictedCategory!! }
        
        val results = mutableListOf<WeekendOverspending>()
        
        // Get historical data for trend analysis
        val historicalStart = LocalDate.now().minusMonths(6)
        val historicalTransactions = allTransactions
            .filter { it.userId == userId }
            .filter { it.amount < 0 }
            .filter { !isInvestmentTransaction(it) }
            .filter { it.date.isAfter(historicalStart) && it.date.isBefore(monthStartDate) }
        
        val historicalByCategory = historicalTransactions
            .filter { it.predictedCategory != null }
            .groupBy { it.predictedCategory!! }
        
        for ((category, categoryTxs) in byCategory) {
            // Separate weekend (Sat, Sun) and weekday (Mon-Fri)
            val weekendTxs = categoryTxs.filter {
                val day = it.date.dayOfWeek
                day == DayOfWeek.SATURDAY || day == DayOfWeek.SUNDAY
            }
            
            val weekdayTxs = categoryTxs.filter {
                val day = it.date.dayOfWeek
                day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY
            }
            
            if (weekendTxs.isEmpty() || weekdayTxs.isEmpty()) {
                continue
            }
            
            val weekendSpending = weekendTxs.sumOf { abs(it.amount) }
            val weekdaySpending = weekdayTxs.sumOf { abs(it.amount) }
            
            // Calculate averages (per transaction)
            val weekendAvg = weekendSpending / weekendTxs.size
            val weekdayAvg = weekdaySpending / weekdayTxs.size
            
            // Calculate ratio
            val ratio = if (weekdayAvg > 0) {
                weekendAvg / weekdayAvg
            } else {
                1.0
            }
            
            // Calculate percentage increase
            val percentageIncrease = if (weekdayAvg > 0) {
                ((weekendAvg - weekdayAvg) / weekdayAvg) * 100
            } else {
                0.0
            }
            
            // Get historical data for trend
            val historicalCategoryTxs = historicalByCategory[category] ?: emptyList()
            val historicalRatios = if (historicalCategoryTxs.isNotEmpty()) {
                val historicalByMonth = historicalCategoryTxs.groupBy { 
                    YearMonth.from(it.date) 
                }
                historicalByMonth.values.map { monthTxs ->
                    val histWeekend = monthTxs.filter { 
                        val day = it.date.dayOfWeek
                        day == DayOfWeek.SATURDAY || day == DayOfWeek.SUNDAY
                    }
                    val histWeekday = monthTxs.filter { 
                        val day = it.date.dayOfWeek
                        day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY
                    }
                    if (histWeekend.isNotEmpty() && histWeekday.isNotEmpty()) {
                        val histWeekendAvg = histWeekend.sumOf { abs(it.amount) } / histWeekend.size
                        val histWeekdayAvg = histWeekday.sumOf { abs(it.amount) } / histWeekday.size
                        if (histWeekdayAvg > 0) histWeekendAvg / histWeekdayAvg else 1.0
                    } else {
                        1.0
                    }
                }
            } else {
                emptyList()
            }
            
            val trend = determineTrend(ratio, historicalRatios)
            val alertLevel = determineAlertLevel(ratio)
            
            val weekendOverspending = WeekendOverspending(
                userId = userId,
                category = category,
                weekendAvg = weekendAvg,
                weekendSpending = weekendSpending,
                weekdayAvg = weekdayAvg,
                weekdaySpending = weekdaySpending,
                ratio = ratio,
                percentageIncrease = percentageIncrease,
                month = currentMonth,
                year = currentMonth.year,
                trend = trend,
                alertLevel = alertLevel,
                isActive = true
            )
            
            results.add(weekendOverspending)
        }
        
        // Save results
        if (results.isNotEmpty()) {
            weekendDao.insertAll(results)
        }
        
        return results
    }
    
    /**
     * Get active weekend overspending for user
     */
    suspend fun getActiveAlerts(userId: Long, month: YearMonth): List<WeekendOverspending> {
        return weekendDao.getActiveByMonthSync(userId, month)
    }
}

